lib/Parse/ParseExpr.cpp

```diff
@@ -3686,15 +3686,18 @@ ParserResult<Expr> Parser::parseExprGradientBody(ExprKind kind) {
     return makeParserCodeCompletionResult<Expr>();
   if (originalFnParseResult.isParseError())
     return errorAndSkipToEnd();
-  // If found comma, parse 'withRespectTo:'.
+  // If found comma, parse 'wrt:'.
   SmallVector<AutoDiffParameter, 8> params;
   if (consumeIf(tok::comma)) {
-    // Parse 'wrt' ':'.
     // If 'withRespectTo' is used, make the user change it to 'wrt'.
     if (Tok.getText() == "withRespectTo") {
-      diagnose(Tok, diag::gradient_expr_use_wrt_not_withrespectto);
+      SourceRange withRespectToRange(Tok.getLoc(), Tok.getRange().getEnd());
+      diagnose(Tok, diag::gradient_expr_use_wrt_not_withrespectto)
+          .highlight(withRespectToRange)
+          .fixItReplace(withRespectToRange, "wrt:");
```

Comment:
I don’t think this is correct. Tok is “withRespectTo” without any colon at the end.

Reply:
Here's what the diagnostic looks like:
```
test.swift:1:16: error: use 'wrt:' to specify parameters to differentiate with respect to
#gradient(foo, withRespectTo: .0)
               ^~~~~~~~~~~~~~
               wrt:
```

Alternatively, using `Tok.getRange().getStart()` instead of `Tok.getRange().getEnd()` yields this:
```
test.swift:1:16: error: use 'wrt:' to specify parameters to differentiate with respect to
#gradient(foo, withRespectTo: .0)
               ^~~~~~~~~~~~~
               wrt
```
I prefer the former, what do you think?

Reply:
When you already have a colon after ‘withRespectTo:’, what will happen? Would your fixit make it ‘wrt::’?

Reply:
Do you mean this?
```
test.swift:1:16: error: use 'wrt:' to specify parameters to differentiate with respect to
#gradient(foo, withRespectTo:: .0)
               ^~~~~~~~~~~~~~
               wrt:
```

This is the current behavior. Would you prefer the following?
```
test.swift:1:16: error: use 'wrt:' to specify parameters to differentiate with respect to
#gradient(foo, withRespectTo:: .0)
               ^~~~~~~~~~~~~
               wrt
```

Reply:
I'm not quite sure what you mean. Would you like a change from the current behavior?
`Tok.getRange().getEnd()` does include the colon, apparently.

Reply:
I'm trying to understand why `Tok.getRange().getEnd()` includes the colon. Shouldn't `Tok` be `withRespectTo`? Why does it include the punctuation at the end?

Reply:
I found the explanation after some research.
`Tok.getRange()` returns a `CharSourceRange`, which is a [half-open character-based source range](https://github.com/apple/swift/blob/8998b27dd465d36caa59535993f11fcf88e4620b/include/swift/Basic/SourceLoc.h#L125).
`CharSourceRange.getEnd()` is the "open end" so it includes an extra character:
```
[test.swift:1:16 - line:1:29] RangeText="withRespectTo:"
[test.swift:1:29 - line:1:30] RangeText=": "
```

On the other hand, `SourceRange` is defined differently ([see here](https://github.com/apple/swift/blob/8998b27dd465d36caa59535993f11fcf88e4620b/include/swift/Basic/SourceLoc.h#L82)):
```
/// SourceRange in swift is a pair of locations.  However, note that *the end
/// location is the start of the last token in the range, not the last character
/// in the range*.  This is unlike SMRange, so we use a distinct type to make
/// sure that proper conversions happen where important.
```

Thus, the correct value for `withRespectToRange` is simply `SourceRange(Tok.getLoc())`.
`SourceRange(Tok.getLoc(), Tok.getRange().getStartLoc())` is the same as `SourceRange(Tok.getLoc(), Tok.getLoc())` and yields the same behavior.

Fixed in https://github.com/apple/swift/pull/17671/commits/3901072c5f4b6c7b7620642eac945b9d5d23d8c0.